internal/pkg/events/events.go

```diff
@@ -47,33 +73,110 @@ func eventusage() {
 }
 
 var (
-	debuge          = func(format string, data ...interface{}) {} // Default no debugging output
-	amqpEventServer *amqp10.AMQPServer
-	serverConfig    saconfig.EventConfiguration
+	debuge       = func(format string, data ...interface{}) {} // Default no debugging output
+	serverConfig saconfig.EventConfiguration
 )
 
-//StartEvents ... entry point to events
+//spawnSignalHandler spawns goroutine which will wait for interruption signal(s)
+// and end smart gateway in case any of the signal is received
+func spawnSignalHandler(watchedSignals ...os.Signal) {
+	interruptChannel := make(chan os.Signal, 1)
+	signal.Notify(interruptChannel, watchedSignals...)
+	go func() {
+		for sig := range interruptChannel {
+			log.Printf("Stopping execution on caught signal: %+v\n", sig)
+			//TO-DO(mmagr): Don't wait based on time, but implement channels to report finished state
+			time.Sleep(2 * time.Second)
+			os.Exit(0)
+		}
+	}()
+}
+
+//spawnAPIServer spawns goroutine which provides http API for alerts and metrics statistics for Prometheus
+func spawnAPIServer(serverConfig saconfig.EventConfiguration, metricHandler *api.EventMetricHandler, amqpHandler *amqp10.AMQPHandler) {
+	prometheus.MustRegister(metricHandler, amqpHandler)
+	// Including these stats kills performance when Prometheus polls with multiple targets
+	prometheus.Unregister(prometheus.NewProcessCollector(prometheus.ProcessCollectorOpts{}))
+	prometheus.Unregister(prometheus.NewGoCollector())
+
+	context := api.NewContext(serverConfig)
+	http.Handle("/alert", api.Handler{Context: context, H: api.AlertHandler})
+	http.Handle("/metrics", promhttp.Handler())
+	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
+		w.Write([]byte(APIHOME))
+	})
+
+	go func() {
+		log.Printf("API server is listening at '%s'\n", serverConfig.API.APIEndpointURL)
+		log.Fatal(http.ListenAndServe(serverConfig.API.APIEndpointURL, nil))
+	}()
+}
+
+//spawnQpidStatusReporter builds gynamic select for reporting status of AMQP connections
+func spawnQpidStatusReporter(applicationHealth *cacheutil.ApplicationHealthCache, qpidStatusCases []reflect.SelectCase) {
+	go func() {
+		for {
+			_, status, _ := reflect.Select(qpidStatusCases)
+			// Note: status here is always very low integer, so we don't need to be afraid of int64>int conversion
+			applicationHealth.QpidRouterState = int(status.Int())
```

Comment:
Maybe I've misread, but won't this set the state to the most recent transition, which may not accurately reflect the global state? I think if you have 2 connections, and both disconnect, this will (correctly) show "0", and then when one reconnects but not the other, it will show "1", suggesting everything is okay when the other connection is still dead.

I wouldn't block this work even if this really is a bug because nothing depends on that value at the moment; but it's definitely something I'd like to test (and have the code explained to me in more detail if I'm wrong!)

Reply:
You did not misread, that is indeed true and this logic was true in the old code. I did not (or at least was trying not to) change the QpidRouterState logic. The only difference is that the state is updated by the all configured sources and not just the one. Overall this needs futher refactor, which will come in next PR.